### PR TITLE
The clear star of yesterday

### PR DESCRIPTION
**Pull request by alexdowad**
Before submitting a PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it)
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [ ] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

**Comment by bbatsov** [comment]
Apart from the build failure, the changes look good.

**Comment by alexdowad** [comment]
Still have to fix up a style violation.

Also profiled autocorrection of some very slow files from Rails. It turns out that `AutocorrectUnlessChangingAST` completely dominated runtime. It is clear that the whole approach of reparsing files to see if a correction is syntactically valid absolutely *kills* performance on large source files.

For autocorrection to be fast, we need to use specific checks to detect cases where autocorrection should not be done. `AutcorrectUnlessChangingAST` is not an option.

I have removed almost all uses of it. The only one left is `Style/Not`.

**Comment by alexdowad** [comment]
Forgot to say: on one large test file from Action Pack which I was using for benchmarking, autocorrection is now more than 10x faster.

File: spec/rubocop/cop/style/braces_around_hash_parameters_spec.rb [hunk 1]
```diff
@@ -160,11 +160,11 @@
     end
 
     context 'in a method call without parentheses' do
-      it 'does not correct a hash parameter with trailing comma' do
+      it 'corrects a hash parameter with trailing comma' do
         # Because `get :i, x: 1,` is invalid syntax.
```
Comment (segiddins) [inline spec/rubocop/cop/style/braces_around_hash_parameters_spec.rb]: This comment seems to suggest that your change produces invalid code?

Comment (alexdowad) [inline spec/rubocop/cop/style/braces_around_hash_parameters_spec.rb]: I'll delete the comment.

**Comment by alexdowad** [comment]
With this change, autocorrection of large and style-error-ridden projects like Rails is quite fast! My stress tests can be run comfortably.

The very first run of the autocorrection stress test has picked up a bunch of bugs. Just looking at a problem where `TrivialAccessors` creates a syntax error.

**Comment by alexdowad** [comment]
Fixed up the broken build.

**Comment by alexdowad** [comment]
Rebased on master.

**Comment by alexdowad** [comment]
Turns out the `TrivialAccessor` problem which I discovered was caused by a bug in `parser`... caused by yours truly. Just sent a patch to `parser`.

**Comment by bbatsov** [comment]
One commit was removed, right?  Please rebase on top of `master`

**Comment by alexdowad** [comment]
And we're back to green.

**Pull request merged**